src/main/java/com/ocient/jdbc/XGConnection.java

```diff
@@ -2283,27 +2403,33 @@ private void clientHandshakeOktaNativeSSO(final String db, final OpenIDAuthentic
 		catch (final Exception e)
 		{
 			LOGGER.log(Level.WARNING, String.format("Exception %s occurred during handshake with message %s", e.toString(), e.getMessage()));
-			// FIXME remove these catch blocks
-			throw e;
-		}
 
-		if (shouldRequestVersion)
-		{
-			if (serverVersion.equals(""))
-			{
-				// FIXME make async
-				blockingGetSQLResult(fetchServerVersion()).get();
-			}
+			OktaNativeSSOCompletionStage = OktaNativeSSOCompletionStage
+				.thenCompose(handshakeSuccessOrEx -> handshakeSuccessOrEx
+				.recoverAsync(ex -> disconnect(false)
+									.thenApply(closeTransportSuccessOrEx -> { 
+
+										if(closeTransportSuccessOrEx.isException())
+											LOGGER.log(Level.WARNING, closeTransportSuccessOrEx.getException(), () -> "Failed to close socket in clientHandshakeOktaNativeSSO");
+
+										return handshakeSuccessOrEx;
+									})));
 		}
-		LOGGER.log(Level.INFO, "Handshake SSO without token finished");
+
+		if (shouldRequestVersion && serverVersion.equals(""))
+			OktaNativeSSOCompletionStage = OktaNativeSSOCompletionStage.thenCompose(successOrEx -> successOrEx.mapSuccessAsync(success -> fetchServerVersion()));
+		
+		return OktaNativeSSOCompletionStage;
 	}
 
-	private void clientHandshakeOktaSessionToken(final String db, final OpenIDAuthenticator authenticator, final boolean shouldRequestVersion) throws Exception {
+	private CompletionStage<SQLResult<Void>> clientHandshakeOktaSessionToken(final String db, final OpenIDAuthenticator authenticator, final boolean shouldRequestVersion) throws Exception {
+
+		CompletionStage<SQLResult<Void>> handshakeOktaSessionCompletionStage = SQLResult.voidFuture();
+
 		try
 		{
 			LOGGER.log(Level.INFO, "Beginning Okta Session Token Exchange");
 
-
 			// This handshake can only be used with the JDBC CLI
 			if (!CLI.isInitialized() || CLI.getReader() == null) {
 				throw new SQLFeatureNotSupportedException(String.format("The JDBC CLI is required to use the %s handshake", HandshakeType.OKTA_NATIVE_SSO));
```

Comment:
If this is the call that requires us to keep the `throws` on the interface, we can give it back to the caller via complete stage
```suggestion
				return SQLResult.failAsync(new SQLFeatureNotSupportedException(String.format("The JDBC CLI is required to use the %s handshake", HandshakeType.OKTA_NATIVE_SSO)));
```